openedx/core/djangoapps/programs/utils.py

```diff
@@ -460,57 +461,86 @@ def _attach_course_run_upgrade_url(self, run_mode):
     def _attach_course_run_may_certify(self, run_mode):
         run_mode['may_certify'] = self.course_overview.may_certify()
 
-    def _check_enrollment_for_user(self, course_run):
-        applicable_seat_types = self.data['applicable_seat_types']
+    def _filter_out_courses_with_entitlements(self, courses):
+        """
+        Removes courses for which the current user already holds an applicable entitlement. 
+        
+        Arguments:
+            courses (list): Containing dicts representing courses in a program
 
-        (enrollment_mode, active) = CourseEnrollment.enrollment_mode_for_user(
-            self.user,
-            CourseKey.from_string(course_run['key'])
-        )
+        Returns:
+            A subset of the given list of course dicts 
+        """
+        course_uuids = set(course['uuid'] for course in courses)
+        # Filter the entitlements' modes with a case-insensitive match against applicable seat_types
+        query = Q()
+        for seat_type in self.data['applicable_seat_types']:
+            query = query | Q(mode__iexact=seat_type)
+        query = Q(course_uuid__in=course_uuids) & query
+        entitlements = self.user.courseentitlement_set.filter(query)
+        courses_with_entitlements = set(str(entitlement.course_uuid) for entitlement in entitlements)
+        return [course for course in courses if course['uuid'] not in courses_with_entitlements]
+
+    def _filter_out_courses_with_enrollments(self, courses):
+        """
+        Removes courses for which the current user already holds an active and applicable enrollment
+        for one of that course's runs.
 
-        is_paid_seat = False
-        if enrollment_mode is not None and active is not None and active is True:
-            # Check all the applicable seat types
-            # this will also check for no-id-professional as professional
-            is_paid_seat = any(seat_type in enrollment_mode for seat_type in applicable_seat_types)
+        Arguments:
+            courses (list): Containing dicts representing courses in a program
 
-        return is_paid_seat
+        Returns:
+            A subset of the given list of course dicts
+        """
+        enrollments = self.user.courseenrollment_set.filter(
+            is_active=True,
+            mode__in=self.data['applicable_seat_types']
```

Comment:
Can we trust `user.mode` to be formatted correctly (not capitalized)?

Reply:
yes because it is populated from the seat product's seat_type attribute, which ensures lowercase as specified above

Reply:
What about `applicable_seat_types`, which seems to come from catalog? Also, do we have to be concerned about encoding (unicode vs str)?

Reply:
Do we need to filter out expired entitlements?

If they purchased the program a year ago but didn't use their entitlement in time, should they see the button and be able to buy it again?

Reply:
That list is populated from the slug field of the seat type model, which will be lower cased.
I am not certain about encoding.

Reply:
This will need to be normalized to account for the fact that the user may have a `no-id-professional` enrollment that is applicable when the program supports `professional` seats.